dm/worker/server.go

```diff
@@ -480,9 +572,54 @@ OUTER:
 	return nil
 }
 
+func (s *Server) handleRelayConfig(ctx context.Context, relayCh chan ha.RelaySource, errCh chan error) error {
+OUTER:
+	for {
+		select {
+		case <-ctx.Done():
+			break OUTER
+		case relaySource, ok := <-relayCh:
+			if !ok {
+				break OUTER
+			}
+			log.L().Info("receive relay source", zap.String("relay source", relaySource.Source), zap.Bool("is deleted", relaySource.IsDeleted))
+			err := s.operateRelaySource(relaySource)
+			s.setSourceStatus(relaySource.Source, err, true)
+			if err != nil {
+				opErrCounter.WithLabelValues(s.cfg.Name, opErrTypeRelaySource).Inc()
+				log.L().Error("fail to operate relay source on worker",
+					zap.String("relay source", relaySource.Source),
+					zap.Bool("is deleted", relaySource.IsDeleted),
+					zap.Error(err))
+				if etcdutil.IsRetryableError(err) {
+					return err
+				}
+			}
+		case err, ok := <-errCh:
+			// currently no value is sent to errCh
+			if !ok {
+				break OUTER
+			}
+			// TODO: Deal with err
+			log.L().Error("WatchRelayConfig received an error", zap.Error(err))
+			if etcdutil.IsRetryableError(err) {
+				return err
+			}
+		}
+	}
+	log.L().Info("worker server is closed, handleRelayConfig will quit now")
+	return nil
+}
+
 func (s *Server) operateSourceBound(bound ha.SourceBound) error {
 	if bound.IsDeleted {
-		return s.stopWorker(bound.Source)
+		// TODO: will worker be modified on other goroutine?
+		w := s.getWorker(true)
```

Comment:
Should we check w==nil? ditto for others.

Reply:
I'll lock the `Server` and remove this TODO also, in next commits